test/native.jl

```diff
@@ -307,6 +307,70 @@ end
     end
 end
 
+using Libdl
+
+function generate_shlib_fptr(f, tt, name=GPUCompiler.safe_name(repr(f)))
+    mktemp() do path, io
+        source = FunctionSpec(f, Base.to_tuple_type(tt), false, name)
+        target = NativeCompilerTarget(;reloc=LLVM.API.LLVMRelocPIC, extern=true)
+        params = TestCompilerParams()
+        job = CompilerJob(target, source, params)
+        obj, _ = GPUCompiler.codegen(:obj, job; strip=true, only_entry=false, validate=false)
+        write(io, obj)
+        flush(io)
+        # FIXME: Be more portable
+        run(`ld -shared -o $path.$dlext $path`)
+        ptr = dlopen("$path.$dlext", Libdl.RTLD_LOCAL)
+        fptr = dlsym(ptr, "julia_$name")
+        @assert fptr != C_NULL
+        atexit(()->rm("$path.$dlext"))
+        fptr
+    end
+end
+
+@static if VERSION >= v"1.7.0-DEV.600"
+@testset "shared library emission" begin
+    f1(x) = x+1
+    @test ccall(generate_shlib_fptr(f1, (Int,)), Int, (Int,), 1) == 2
+    f2(x,y) = x+y
+    @test ccall(generate_shlib_fptr(f2, (Int,Int)), Int, (Int,Int), 1, 2) == 3
+    f3(str) = str*"!"
+    @test_skip ccall(generate_shlib_fptr(f3, (String,)), String, (String,), "Hello") == "Hello!"
+    function f4()
+        # Something reasonably complicated
+        if isdir(homedir())
+            true
+        else
+            false
+        end
+    end
+    @test ccall(generate_shlib_fptr(f4, ()), Bool, ())
+    f5() = :asymbol
+    @test_skip ccall(generate_shlib_fptr(f5, ()), Symbol, ()) == :asymbol
+    f6(x) = x == :asymbol ? true : false
+    @test_skip ccall(generate_shlib_fptr(f6, (Symbol,)), Bool, (Symbol,), :asymbol)
+    @test_skip !ccall(generate_shlib_fptr(f6, (Symbol,)), Bool, (Symbol,), :bsymbol)
```

Comment:
At least one of these test should run in an external process.

Reply:
In reality, anything that relies on runtime calls/state should probably go through another process. I'll spool one up with Distributed and see how far I can get.

Reply:
This approach was very enlightening, if also unfortunate. I found that we still emit things like Symbols and some other objects as `inttoptr`'d pointers. I'm going to have one of the simpler tests run on another process, but clearly there's still some extra work to be done.